File: src/main/java/com/github/jactor/persistence/time/Now.java
package com.github.jactor.persistence.time;

import java.time.LocalDateTime;

public class Now {
    private static final Object SYNC = new Object();

    private static volatile Now instance;

    protected LocalDateTime nowAsDateTime() {
        return LocalDateTime.now();
    }

    public static LocalDateTime asDateTime() {
        return instance.nowAsDateTime();
    }

    static void reset(Now instance) {
        synchronized (SYNC) {
            Now.instance = instance;
        }
    }

    static {
        reset(new Now());
    }
}

File: src/main/kotlin/com/github/jactor/persistence/time/Now.kt
package com.github.jactor.persistence.time

import java.time.LocalDateTime

open class Now {
    protected open fun nowAsDateTime(): LocalDateTime {
        return LocalDateTime.now()
    }

    companion object {
        private val SYNC = Any()

        @Volatile
        private var instance: Now? = null

        @JvmStatic
        fun asDateTime(): LocalDateTime {
            return instance!!.nowAsDateTime()
        }

        @JvmStatic
        fun reset(instance: Now?) {
            synchronized(SYNC) { Companion.instance = instance }
        }

        init {
            reset(Now())
        }
    }
}
